File: tst/Kroki.Tests/Output/OneForm.cs
using System;
using System.Collections.Generic;
using System.Text;
using Kroki.Runtime;
using TD3DXMatrix = TD3DXVector3;
using TD3DXQuaternion = TD3DXVector2;
using TInt64a = Int64;
using TInt64b = Comp;
using TInt64F = TInt64b;
using FInt64 = TInt64F;
using Int64D = TInt64a;

namespace Kroki.Example
{
    public struct PD3DXVector2
    {
    }

    public struct TD3DXVector2
    {
        public float x;
        public float y;
    }

    public struct TD3DXVector3
    {
        public float x;
        public float y;
        public float z;
    }

    public interface IsMatrix
    {
        TD3DXMatrix D3DXMatrix(float m00, float m01, float m02, float m03, float m10, float m11, float m12, float m13, float m20, float m21, float m22, float m23, float m30, float m31, float m32, float m33);
        TD3DXMatrix D3DXMatrixAdd(TD3DXMatrix m1, TD3DXMatrix m2, TD3DXMatrix mOut);
        TD3DXMatrix D3DXMatrixSubtract(TD3DXMatrix m1, TD3DXMatrix m2, TD3DXMatrix mOut);
        TD3DXMatrix D3DXMatrixMul(TD3DXMatrix m, float MulBy, TD3DXMatrix mOut);
        bool D3DXMatrixEqual(TD3DXMatrix m1, TD3DXMatrix m2);
        TD3DXQuaternion D3DXQuaternionConjugate(TD3DXQuaternion qOut, TD3DXQuaternion q);
    }

    public struct PID3DXBuffer
    {
    }

    public interface ID3DXBuffer
    {
        IntPtr GetBufferPointer();
        ulong GetBufferSize();
    }

    public class TOneForm : TForm
    {
        public TButton LoginBtn;
        public TEdit NameBox;
        public void LoginBtnClick(object Sender)
        {
            Console.WriteLine("Hello!");
        }

        public static readonly TD3DXVector2 D3DXVector2Zero = new()
        {
            x = 0,
            y = 0
        };
        public static readonly TD3DXVector3 D3DXVector3Zero = new()
        {
            x = 0,
            y = 0,
            z = 0
        };
        public bool D3DXVector2Equal(TD3DXVector2 v1, TD3DXVector2 v2)
        {
            Console.WriteLine((v1), (v2));
            return true;
        }
    }

    public struct PCardinal
    {
    }

    public struct PInt64
    {
    }

    public class AInt64F : List<TInt64F>
    {
    }

    public struct PAInt64F
    {
    }

    public class TPERF_COUNTER_BLOCK
    {
        public uint ByteLength;
    }

    public struct PPERF_COUNTER_BLOCK
    {
    }

    public enum TDrawFlag
    {
        dfDontDraw,
        dfWordWrap,
        dfStretchImages,
        dfAlignMiddle,
        dfAlignBottom
    }

    public class TDrawFlags : List<TDrawFlag>
    {
    }

    public enum TPartType
    {
        ptNone,
        ptText,
        ptColor,
        ptStyle,
        ptBitmap
    }

    public enum TFormBorderStyle
    {
        bsNone,
        bsSingle,
        bsSizeable,
        bsDialog,
        bsToolWindow,
        bsSizeToolWin
    }

    public static class OneForm
    {
        public static Range TBorderStyle = bsNone..bsSingle;
        public static Range TBorderSizes = 23..42;
        public static Range TBorderSigns = "A".."z";
        public static bool ParseThat(char c)
        {
            string tmp = default;
            tmp = (c);
            Console.WriteLine(tmp);
            return c.In(new[] { "a".."z", " ", "A".."Z" });
        }
    }

    public delegate void TOnLinkClick(object Sender, string Link);
    public enum TStickText
    {
        stTop,
        stNone,
        stBottom
    }

    public interface IDirectXFile
    {
        void Draw();
    }
}

File: tst/Kroki.Tests/Input/OneForm.pas
unit OneForm;

{$mode ObjFPC}{$H+}

interface

uses
  Classes, SysUtils, Forms, Controls, Graphics, Dialogs, StdCtrls;

const
    Processor_IDX_Str = '238';
    Processor_IDX = 238;
    CPUUsageIDX = 6;
    d3dx8dll = 'D3DX8ab.dll';
    D3DX_PI    : Single  = 3.141592654;
    D3DX_1BYPI : Single  = 0.318309886;
    D3DX_DEFAULT : Cardinal = $FFFFFFFF;

type
  PD3DXVector2 = ^TD3DXVector2;
  TD3DXVector2 = packed record
    x, y : Single;
  end;
  TD3DXVector3 = packed record
    x, y, z : Single;
  end;

  TD3DXMatrix = TD3DXVector3;
  TD3DXQuaternion =        TD3DXVector2;
  IsMatrix = interface
    function D3DXMatrix(const m00, m01, m02, m03,
                          m10, m11, m12, m13,
                          m20, m21, m22, m23,
                          m30, m31, m32, m33 : Single) : TD3DXMatrix;
    function D3DXMatrixAdd(const m1,m2 : TD3DXMatrix; out mOut : TD3DXMatrix) : TD3DXMatrix;
    function D3DXMatrixSubtract(const m1, m2 : TD3DXMatrix; out mOut : TD3DXMatrix) : TD3DXMatrix;
    function D3DXMatrixMul(const m : TD3DXMatrix; MulBy : Single; out mOut : TD3DXMatrix) : TD3DXMatrix;
    function D3DXMatrixEqual(const m1, m2 : TD3DXMatrix) : Boolean;
    function D3DXQuaternionConjugate(out qOut : TD3DXQuaternion; const q : TD3DXQuaternion) : TD3DXQuaternion;
  end;

  PID3DXBuffer = ^ID3DXBuffer;
  ID3DXBuffer = interface(IUnknown)
    function GetBufferPointer : Pointer; stdcall;
    function GetBufferSize : LongWord; stdcall;
  end;

  { TOneForm }

  TOneForm = class(TForm)
    LoginBtn: TButton;
    NameBox: TEdit;
    procedure LoginBtnClick(Sender: TObject);
  private

  public
    const D3DXVector2Zero : TD3DXVector2 = (x : 0; y : 0);
    const D3DXVector3Zero : TD3DXVector3 = (x : 0; y : 0; z : 0);
    function D3DXVector2Equal(const v1, v2 : TD3DXVector2) : Boolean;
  end;

  PCardinal = ^Cardinal;

  TInt64a = Int64;
  TInt64b = Comp;
  PInt64 = ^TInt64a;

  TInt64F = TInt64b;
  FInt64 = TInt64F;
  Int64D = TInt64a;

  AInt64F = array[0..$FFFF] of TInt64F;
  PAInt64F = ^AInt64F;

  TPERF_COUNTER_BLOCK = record
    ByteLength : DWORD;
  end;
  PPERF_COUNTER_BLOCK = ^TPERF_COUNTER_BLOCK;

  TDrawFlag = (dfDontDraw, dfWordWrap, dfStretchImages, dfAlignMiddle, dfAlignBottom);
  TDrawFlags = set of TDrawFlag;

  TPartType = (ptNone, ptText, ptColor, ptStyle, ptBitmap);

  TFormBorderStyle = (bsNone, bsSingle, bsSizeable, bsDialog, bsToolWindow, bsSizeToolWin);

  TBorderStyle = bsNone..bsSingle;
  TBorderSizes = 23..42;
  TBorderSigns = 'A'..'z';

  TOnLinkClick = procedure(Sender: TObject; Link: String) of object;

  TStickText = (stTop, stNone, stBottom);

  IDirectXFile = interface
      procedure Draw;
  end;

var
  OneForm2: TOneForm;
  DirectXFileCreate : function (out lplpDirectXFile: IDirectXFile) : HResult; stdcall;

implementation

{$R *.lfm}

{ TOneForm }

procedure TOneForm.LoginBtnClick(Sender: TObject);
begin
  WriteLn('Hello!');
end;

function TOneForm.D3DXVector2Equal(const v1, v2: TD3DXVector2): Boolean;
begin
  WriteLn(string(v1), string(v2));
  D3DXVector2Equal := True;
end;

function ParseThat(c: char) : Boolean;
var
  tmp: string;
begin
  tmp := string(c);
  WriteLn( tmp );
  Result := c in ['a'..'z', ' ', 'A'..'Z'];
end;

end.

